src/hooks/click-outside.js

```diff
@@ -0,0 +1,17 @@
+import { useEffect } from 'react'
+
+export function useClickOutside(el, onClick) {
+  const els = [].concat(el)
+
+  const handler = (ev) => {
+    const target = ev.target
+    if (els.every((ref) => !ref.current.contains(target))) {
+      onClick(ev)
```

Comment:
Maybe add the `window.removeEventListener('click', handler)` inside here also to clean it up on off click but not necessarily unmount 👍

Reply:
Won't this remove the handler for that component on the first off-click? I think we want to keep it there until it's unmounted.

I guess we do need to check each `ref` exists though, as each may be unmounted, and the cleanup handler of `useEffect` might not be called it may not be a ref to the component that is consuming the hook. (If that makes sense).

I guess the handler should also be removed if all `els` are falsey.

Reply:
> I guess the handler should also be removed if all `els` are falsey.

Actually, probably shouldn't do that. There is nothing stopping a component passing an array of refs to components _other_ than itself, and this component will expect the handler to stay around for _it's own_ lifetime, even if all the targets are unmounted and remounted again. I'll go with the truthy check for now.

It's a bit of an edge case I guess, I imagine most usage will be the component passing its own ref.

Reply:
It will remove it you're right. And that's because the effect only runs when `el` changes.
This does bring up something I think we've overlooked though 🤦‍♂️ 
You don't want to bind the handler on every render. 
Think of the case where you have a menu that slides in and out. You don't want the handler firing every time the user clicks the page even if the menu is closed. So maybe it requires a parameter that tells it if it can bind, then we can conditionally bind the event in the `useEffect` function.

Reply:
> You don't want to bind the handler on every render

I _thought_ that it bound only when `el` changed, (as well as mount/unmount) as `el`  is passed in the array of dependencies (last argument to `useEffect`). Am I wrong there? 

In any case, I think we _only_ want to bind the handler on mount, and remove on unmount, so I think I can pass an empty array as last argument to achieve that.

> Passing in an empty array [] of inputs tells React that your effect doesn’t depend on any values from the component, so that effect would run only on mount and clean up on unmount; it won’t run on updates. - https://reactjs.org/docs/hooks-reference.html#useeffect

Reply:
Yes. Sorry I probably didn't word that right.

Currently, that handler will fire on every click as soon as the component is mounted. You don't want that.

You only want to bind when some condition is met. For example, a menu in an open state.

Reply:
That's an interesting one... should be enforce that this condition be supplied, or would you be OK with opt-in, like https://github.com/nearform/react-browser-hooks/pull/100/commits/09b7b01727b01945e8400bf821cd6c14b8296e1e?

Reply:
Yeah I think that makes sense 👍